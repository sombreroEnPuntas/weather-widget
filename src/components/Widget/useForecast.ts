import { useEffect, useState } from 'react'
import { useRouter } from 'next/router'
import { HourlyData } from '../../apiClient/forecast'

const useForecast = () => {
  const router = useRouter()
  const [data, setData] = useState<HourlyData>()
  const [error, setError] = useState<string>()

  useEffect(() => {
    ;(async () => {
      try {
        const response = await fetch(router.basePath + `/api/forecast`)
        const result = await response.json()

        setData(result.data)
        setError(result.error)
      } catch (e) {
        setData(undefined)
        setError('Server error. Try again later.')
      }
    })()
  }, [])

  return { data, error }
}

export default useForecast
