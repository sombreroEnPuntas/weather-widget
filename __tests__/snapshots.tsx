import React from 'react'
import { render, fireEvent } from '@testing-library/react'

// pages
import Index from '../pages/index'

// deps
import useForecast from '../src/components/Widget/useForecast'

// utils
import {
  forecastByDateFailureMock,
  forecastByDateSuccessMock,
} from '../src/utils/mocks'

// Mocks
jest.mock('../src/components/Widget/useForecast')

// Mock data
const setMock = ({ forecastData }) => {
  ;(useForecast as jest.Mock).mockImplementation(() => forecastData)
}

describe.each`
  Case         | forecastData
  ${'error'}   | ${forecastByDateFailureMock}
  ${'data'}    | ${forecastByDateSuccessMock}
  ${'loading'} | ${{}}
`('Side effect: $Case', ({ forecastData }) => {
  test('snapshot', () => {
    setMock({ forecastData })
    const { container } = render(<Index />)

    expect(container).toMatchSnapshot()
  })
})

describe('Interactions', () => {
  test('reacts to choosing an element', () => {
    setMock({ forecastData: forecastByDateSuccessMock })
    const { getByTestId } = render(<Index />)
    expect(getByTestId('articleOverlineBig')).toHaveTextContent('13°')

    const button = getByTestId('2020-09-29-12:00')
    fireEvent.click(button)

    expect(getByTestId('articleOverlineBig')).toHaveTextContent('12°')
  })
})
