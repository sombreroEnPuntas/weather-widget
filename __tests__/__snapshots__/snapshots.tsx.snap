// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Side effect: data snapshot 1`] = `
<div>
  <h3>
    weather-widget
  </h3>
  <pre>
    <a
      href="https://github.com/sombreroEnPuntas/weather-widget"
    >
      Check the code on github!
    </a>
  </pre>
  <div
    class="wrapper"
  >
    <header
      class="header"
    >
      <article
        class="article"
      >
        <img
          alt="cloudy sky"
          class="image"
          height="100%"
          src="/weather-cloud.svg"
          width="100%"
        />
        <div>
          <div
            class="articleHeading"
          >
            Cloudy
               
            13
            °
             / 
            12
            °
          </div>
          <div
            class="articleOverlineBig"
            data-testid="articleOverlineBig"
          >
            13
            °
          </div>
        </div>
      </article>
      <article
        class="article"
      >
        <div>
          <div
            class="articleHeading"
          >
            Rivendell
          </div>
          <div
            class="articleOverline"
          >
            <span
              style="white-space: pre-line;"
            >
              Donnerstag
 16. Februar
            </span>
          </div>
        </div>
      </article>
    </header>
    <ul
      class="hourlyData indiana-scroll-container indiana-scroll-container--hide-scrollbars"
    >
      <li
        class="hourlyDataItem"
      >
        <button
          class="button"
          data-testid="2020-09-29-11:00"
        >
          <div
            class="title"
          >
            11:00
          </div>
          <img
            alt="cloudy sky"
            class="image"
            height="100%"
            src="/weather-cloud.svg"
            width="100%"
          />
          <div
            class="content"
          >
            13
            °
          </div>
        </button>
      </li>
      <li
        class="hourlyDataItem"
      >
        <button
          class="button"
          data-testid="2020-09-29-12:00"
        >
          <div
            class="title"
          >
            12:00
          </div>
          <img
            alt="clear sky"
            class="image"
            height="100%"
            src="/weather-sun.svg"
            width="100%"
          />
          <div
            class="content"
          >
            12
            °
          </div>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Side effect: error snapshot 1`] = `
<div>
  <h3>
    weather-widget
  </h3>
  <pre>
    <a
      href="https://github.com/sombreroEnPuntas/weather-widget"
    >
      Check the code on github!
    </a>
  </pre>
  <div
    class="wrapper"
  >
    418: The resulting entity body MAY be short and stout.
  </div>
</div>
`;

exports[`Side effect: loading snapshot 1`] = `
<div>
  <h3>
    weather-widget
  </h3>
  <pre>
    <a
      href="https://github.com/sombreroEnPuntas/weather-widget"
    >
      Check the code on github!
    </a>
  </pre>
  <div
    class="wrapper"
  >
    loading...
  </div>
</div>
`;
